sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+
+      // First string is the path to executable, and not the actual first argument.
+      {
+        std::string const Shell = "/bin/sh";
+        EnsureShellExists(Shell);
+        AppendToArgvValues(m_argvValues, argvValuePositions, Shell);
+      }
+
+      // Second argument is the shell switch that tells the command line shell to execute a command
+      AppendToArgvValues(m_argvValues, argvValuePositions, "-c");
+
+      // Third value is the command that needs to be executed.
+      AppendToArgvValues(m_argvValues, argvValuePositions, command);
+
+      // We are done appending to m_argvValues, so it is now safe to grab addresses to the elements
+      // in it.
+      for (auto const pos : argvValuePositions)
+      {
+        m_argv.push_back(m_argvValues.data() + pos);
+      }
+    }
+
+    // argv last element needs to be NULL pointer.
+    m_argv.push_back(NULL);
+  }
+
+  // Form the 'envp' array:
+  // * An array of pointers to non-const C strings (0-terminated).
+  // * Strings are in form key=value (PATH uses ':' as separator)
+  // * Last element is NULL pointer.
+  // * First element (at index 0) is path to a program.
+  {
+    auto const actualPathVarValue = Environment::GetVariable("PATH");
+    auto const processPathVarStatement = std::string("PATH=") + actualPathVarValue
+        + (actualPathVarValue.empty() ? "" : ":") + "/usr/bin:/usr/local/bin";
+
+    m_envpValues.insert(
+        m_envpValues.end(), processPathVarStatement.begin(), processPathVarStatement.end());
+
+    m_envpValues.push_back('\0');
+
+    // We should only grab m_envpValues.data() as we're done appending to it, because appends may
+    // reallocate the buffer to a different memory location.
+    m_envp.push_back(m_envpValues.data());
+    m_envp.push_back(NULL);
+  }
+
+  // Set up pipe communication for the process.
+  posix_spawn_file_actions_init(&m_actions);
+  posix_spawn_file_actions_addclose(&m_actions, outputPipe.m_fd[0]);
+  posix_spawn_file_actions_adddup2(&m_actions, outputPipe.m_fd[1], 1);
+  posix_spawn_file_actions_addclose(&m_actions, outputPipe.m_fd[1]);
+
+  {
+    auto const spawnResult
+        = posix_spawn(&m_pid, m_argv[0], &m_actions, NULL, m_argv.data(), m_envp.data());
+
+    // LCOV_EXCL_START
+    if (spawnResult != 0)
+    {
+      m_pid = -1;
+      this->~ShellProcess();
+      ThrowIfApiCallFails(spawnResult, "Cannot spawn process");
+    }
+    // LCOV_EXCL_STOP
+  }
+
+  close(outputPipe.m_fd[1]);
+  outputPipe.m_fd[1] = -1;
+#endif
+}
+
+ShellProcess::~ShellProcess()
+{
+#if defined(AZ_PLATFORM_WINDOWS)
+  static_cast<void>(CloseHandle(m_processHandle));
+#else
+  if (m_pid > 0)
+  {
+    static_cast<void>(waitpid(m_pid, NULL, 0));
+  }
+
+  posix_spawn_file_actions_destroy(&m_actions);
+#endif
+}
+
+void ShellProcess::Terminate()
+{
+#if defined(AZ_PLATFORM_WINDOWS)
+  static_cast<void>(TerminateProcess(m_processHandle, 0));
+#else
+  if (m_pid > 0)
+  {
+    static_cast<void>(kill(m_pid, SIGKILL));
```

Comment:
There's a window here where m_pid could be re-used between when you decided to terminate the shell and when you actually killed the process. I don't know if that matters, but it *is* a possibility.

Reply:
There doesn't seem to be any atomic way to kill a process. But FWIW, I think OSes try not to reuse PIDs until wrap-around happens. Some even (OpenBSD?) don't give them sequentially so you can't even exploit the fact of guessing how many processes there are, what is the system uptime, or when the wrap-around will happen.

Reply:
This race condition is exactly why CreateProcess returns a handle to the process - it removes the race condition. But the window exists whenever you use process ID to kill/terminate a process.